### PR TITLE
Fix getRegistrationByAdress (use index and handle address conflict)

### PR DESCRIPTION
**Pull request by sbernard31**
See #464 for more details.

### CODE REVIEW

**Comment by AlexITC** [comment]
Based on the question, what I see is that we are discarding registrations having the same ip address, shouldn't this cause problems with clients behind the same NAT?

**Comment by sbernard31** [comment]
What do you mean "we are discarding registrations having the same ip address" ? 

Registration was not discarded but the index (addr => registration) keeps the most recent registration.

**Comment by AlexITC** [comment]
From what I see this doesn't not resolve the actual question, in my opinion, the ideal scenario would be to return a list of registrations instead of a single one.

**Comment by sbernard31** [comment]
Ok, let me resume that, in a leshan server context : 

[PskStore.getIdentity(InetSocketAddress inetAddress)](https://github.com/eclipse/californium/blob/2.0.x/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/pskstore/PskStore.java#L76) is used when the server want to initiate a DTLS connection. This should return only 1 identity, this make sense as we want to talk to 1 specific device/client.

In Leshan the way to know the psk identity, is to search in RegistrationStore to find which registration is associated to this socket address. 

Theoretical there is only 1 device for 1 address, but with NAT a device D1 could get an address A then go to sleep, then a new device D2 will get this free address A. So in the store 2 devices with the same address A. But in fact there is only 1 device behind the A address :D2. D1 lost this address and we don't know its new address, we will know that next time it will update its registration.

So this makes senses to me to only keep tracks to the device which use the address A lastly.

> From what I see this doesn't not resolve the actual question,

I'm not sure to understand what is the question ? :)

>  the ideal scenario would be to return a list of registrations instead of a single one.

I do not understand what does it bring more, as pskStore must choose one of them to choose 1 identity.

**Comment by AlexITC** [comment]
Sounds reasonable, thanks!

File: leshan-server-cluster/src/main/java/org/eclipse/leshan/server/cluster/RedisRegistrationStore.java [hunk 1]
```diff
@@ -139,15 +140,20 @@ public Deregistration addRegistration(Registration registration) {
...
+                // add registration: secondary indexes
+                byte[] regid_idx = toRegIdKey(registration.getId());
+                j.set(regid_idx, registration.getEndpoint().getBytes(UTF_8));
+                byte[] addr_idx = toRegAddrKey(registration.getSocketAddress());
+                j.set(addr_idx, registration.getEndpoint().getBytes(UTF_8));
 
                 if (old != null) {
                     Registration oldRegistration = deserializeReg(old);
                     // remove old secondary index
                     if (registration.getId() != oldRegistration.getId())
```
Comment (msangoi) [inline leshan-server-cluster/src/main/java/org/eclipse/leshan/server/cluster/RedisRegistrationStore.java]: Not related to this PR but this string equality looks buggy...

Comment (sbernard31) [reply leshan-server-cluster/src/main/java/org/eclipse/leshan/server/cluster/RedisRegistrationStore.java]: I will fix this in a separated commit

**Pull request merged**